nanopb: build fails when used in external module

**Describe the bug**

The documentation describes how to use nanopb:

https://docs.zephyrproject.org/latest/services/serialization/nanopb.html

This works fine as long as it is used in the main CMakeLists.txt. If the same procedure is used in an external module the build fails with 

```

CMake Error at C:/Users/KleiserS/Desktop/test/example/deps/modules/lib/nanopb/extra/FindNanopb.cmake:407 (target_compile_features):
  target_compile_features no known features for C compiler

  "GNU"

  version 12.2.0.

```

This happend in general before in #70516, was reverted and fixed later in #70658. I am not sure if it is fixed completely, because my module depending on nanopb fails now.

Commit 6b27cb2b835cd817c191ae9fe1471065c2693d1a is the last one working.

**To Reproduce**
I have prepared an example project. Steps to reproduce this behavior:
1. git clone https://github.com/sageve/nanopb-example
2. cd nanopb-example
3. west init -l
4. west update
5. west build -b qemu_cortex_a9 or any other board

**Expected behavior**
The build should not fail. In case I am misusing nanopb, it should be documented how to use it in modules.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -219,6 +219,26 @@ if(CONFIG_LTO)
   add_link_options($<TARGET_PROPERTY:linker,lto_arguments>)
 endif()
 
+if(CONFIG_STD_C23)
+  set(CSTD c2x)
+elseif(CONFIG_STD_C17)
+  set(CSTD c17)
+elseif(CONFIG_STD_C11)
+  set(CSTD c11)
+elseif(CONFIG_STD_C99)
+  set(CSTD c99)
+elseif(CONFIG_STD_C90)
+  set(CSTD c90)
+else()
+  message(FATAL_ERROR "Unreachable code. Expected C standard to have been chosen.")
+endif()
+
+if(CONFIG_GNU_C_EXTENSIONS)
+  string(REPLACE "c" "gnu" CSTD "${CSTD}")
+endif()
+
+list(APPEND CMAKE_C_COMPILE_FEATURES ${compile_features_${CSTD}})
+
 # @Intent: Obtain compiler specific flags related to C++ that are not influenced by kconfig
 zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler-cpp,required>>)
 
@@ -956,35 +976,19 @@ if(CONFIG_USERSPACE)
   set(PROCESS_GPERF ${ZEPHYR_BASE}/scripts/build/process_gperf.py)
 endif()
 
-get_property(CSTD GLOBAL PROPERTY CSTD)
-if(NOT DEFINED CSTD)
-  if(CONFIG_STD_C23)
-    set(CSTD c2x)
-  elseif(CONFIG_STD_C17)
-    set(CSTD c17)
-  elseif(CONFIG_STD_C11)
-    set(CSTD c11)
-  elseif(CONFIG_STD_C99)
-    set(CSTD c99)
-  elseif(CONFIG_STD_C90)
-    set(CSTD c90)
-  else()
-    message(FATAL_ERROR "Unreachable code. Expected C standard to have been chosen.")
-  endif()
-
-  if(CONFIG_GNU_C_EXTENSIONS)
-    string(REPLACE "c" "gnu" CSTD "${CSTD}")
-  endif()
-else()
+get_property(GLOBAL_CSTD GLOBAL PROPERTY CSTD)
+if(DEFINED GLOBAL_CSTD)
   message(DEPRECATION
     "Global CSTD property is deprecated, see Kconfig.zephyr for C Standard options.")
+  set(CSTD ${GLOBAL_CSTD})
+  list(APPEND CMAKE_C_COMPILE_FEATURES ${compile_features_${CSTD}})
 endif()
 
 # @Intent: Obtain compiler specific flag for specifying the c standard
 zephyr_compile_options(
   $<$<COMPILE_LANGUAGE:C>:$<TARGET_PROPERTY:compiler,cstd>${CSTD}>
 )
-set(CMAKE_C_COMPILE_FEATURES ${compile_features_${CSTD}} PARENT_SCOPE)
+set(CMAKE_C_COMPILE_FEATURES ${CMAKE_C_COMPILE_FEATURES} PARENT_SCOPE)
 
 # @Intent: Configure linker scripts, i.e. generate linker scripts with variables substituted
 toolchain_ld_configure_files()

